File: src/tp04/ej09/BankAccount.java
package tp04.ej09;

public abstract class BankAccount {
    private int id;
    private double balance;

    public BankAccount(int id) {
        this.id = id;
        this.balance = 0;
    }

    public double getBalance() {
        return this.balance;
    }

    protected abstract boolean canExtract(double amount);

    public void extract(double amount) {
        if (!canExtract(amount)) throw new RuntimeException("No cuenta con los fondos necesarios.");
        this.balance -= amount;
    }

    public void deposit(double amount) {
        this.balance += amount;
    }

    public String toString() {
        return "ID: " + this.id + " Balance: " + this.balance;
    }
}

File: src/tp08/ej05/bank_account.rb
class BankAccount

 def initialize
   raise 'Debe sobreescribir este método'
 end

 def create(id)
   @id = id
   @balance = 0
 end

 def deposit(amount)
   @balance += amount
 end

 def extract(amount)
   raise 'No cuenta con los fondos necesarios' unless extract? amount
   @balance -= amount
 end

 def extract?(_amount)
   raise 'Debe sobreescribir este método'
 end

 def to_s
   "Cuenta #{@id} con saldo #{@balance}"
 end

 private :create, :extract?

end
